### PR TITLE
[Feature] Support MultiScaleDeformableAttn with cambricon MLU backend

### PR DESCRIPTION
**Pull request by ZShaopeng**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.
 
## Motivation
 
The motivation of the PR is to support running MultiScaleDeformAttn with Cambricon MLU backend.
 
It includes five parts:
 
    1. Add ms_deform_attn_mlu_kernel.mlu src code as MLU kernel.
    2. Add ms_deform_attn_mlu.cpp to support launching kernel in PyTorch.
    3. Modify common_mlu_helper.hpp to add common function.
    4. Refactor multi_scale_deform_attn.py to support MLU backend.
    5. Refactor test_ms_deformable_attn.py to support testing MultiScaleDeformAttn with MLU backend.
 
## Modification
 
- MLU src code
  Add MLU src code of MultiScaleDeformAttn in directory:
  mmcv/ops/csrc/common/mlu/ms_deform_attn_mlu_kernel.mlu,
  mmcv/ops/csrc/common/mlu/common_mlu_helper.hpp.
- PyTorch adaptation
  Adapt MultiScaleDeformAttn for PyTorch in directory:
  mmcv/ops/csrc/pytorch/mlu/ms_deform_attn_mlu.cpp,
  mmcv/ops/multi_scale_deform_attn.py.
- Unit test
  Support test MultiScaleDeformAttn with various backends in test_ms_deformable_attn.py.
 
## BC-breaking (Optional)
 
Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.
 
## Use cases (Optional)
 
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.
 
## Checklist
 
**Before PR**:
 
- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.
 
**After PR**:
 
- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tests/test_ops/test_ms_deformable_attn.py [hunk 1]
```diff
@@ -102,40 +107,66 @@ def test_forward_equal_with_pytorch_double():
...
+])
+@pytest.mark.parametrize('dtype', [
+    torch.float,
+    pytest.param(
+        torch.double,
+        marks=pytest.mark.skipif(
+            IS_MLU_AVAILABLE,
+            reason='MLU does not support for 64-bit floating point')),
+    torch.half
+])
```
Comment (zhouzaida) [inline tests/test_ops/test_ms_deformable_attn.py:169]:
```suggestion
@pytest.mark.parametrize('device', [
    pytest.param(
        'cuda',
        marks=pytest.mark.skipif(
            not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
    pytest.param(
        'mlu',
        marks=pytest.mark.skipif(
            not IS_MLU_AVAILABLE, reason='requires MLU support'))
])
@pytest.mark.parametrize('dtype', [
    torch.float,
    pytest.param(
        torch.double,
        marks=pytest.mark.skipif(
            IS_MLU_AVAILABLE,
            reason='MLU does not support for 64-bit floating point')),
    torch.half
])
```

File: tests/test_ops/test_ms_deformable_attn.py [hunk 2]
```diff
@@ -145,20 +176,23 @@ def test_forward_equal_with_pytorch_float():
     1025,
 ])
 def test_gradient_numerical(channels,
+                            device_type,
```
Comment (zhouzaida) [inline tests/test_ops/test_ms_deformable_attn.py:179]:
```suggestion
                            device,
```

File: tests/test_ops/test_ms_deformable_attn.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 from mmcv.ops.multi_scale_deform_attn import (
     MultiScaleDeformableAttention, MultiScaleDeformableAttnFunction,
     multi_scale_deformable_attn_pytorch)
+from mmcv.utils import IS_MLU_AVAILABLE
```
Comment (zhouzaida) [inline tests/test_ops/test_ms_deformable_attn.py:8]:
```suggestion
from mmcv.utils import IS_CUDA_AVAILABLE, IS_MLU_AVAILABLE
```

File: tests/test_ops/test_ms_deformable_attn.py [hunk 4]
```diff
@@ -102,40 +107,66 @@ def test_forward_equal_with_pytorch_double():
...
 
 
-@pytest.mark.skipif(
-    not torch.cuda.is_available(), reason='requires CUDA support')
-def test_forward_equal_with_pytorch_float():
+@pytest.mark.parametrize('device_type', [
+    pytest.param(
+        'cuda',
+        marks=pytest.mark.skipif(
+            not torch.cuda.is_available(), reason='requires CUDA support')),
```
Comment (zhouzaida) [inline tests/test_ops/test_ms_deformable_attn.py:114]:
```suggestion
            not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
```

File: mmcv/ops/multi_scale_deform_attn.py [hunk 5]
```diff
@@ -351,6 +352,10 @@ def forward(self,
             output = MultiScaleDeformableAttnFunction.apply(
                 value, spatial_shapes, level_start_index, sampling_locations,
                 attention_weights, self.im2col_step)
+        elif IS_MLU_AVAILABLE and value.is_mlu:
+            output = MultiScaleDeformableAttnFunction.apply(
+                value, spatial_shapes, level_start_index, sampling_locations,
+                attention_weights, self.im2col_step)
```
Comment (zhouzaida) [inline mmcv/ops/multi_scale_deform_attn.py:358]:
```suggestion
        if (IS_CUDA_AVAILABLE and value.is_cuda or
            IS_MLU_AVAILABLE and value.is_mlu):
            output = MultiScaleDeformableAttnFunction.apply(
                value, spatial_shapes, level_start_index, sampling_locations,
                attention_weights, self.im2col_step)
```

**Comment by grimoire** [comment]
Unit test `test_gradient_numerical` gradcheck failed.

**Review by grimoire — APPROVED** [review]
LGTM

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**